src/python_testing/TC_TSTAT_2_2.py

```diff
@@ -0,0 +1,778 @@
+#
+#    Copyright (c) 2024 Project CHIP Authors
+#    All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+# See https://github.com/project-chip/connectedhomeip/blob/master/docs/testing/python.md#defining-the-ci-test-arguments
+# for details about the block below.
+#
+# === BEGIN CI TEST ARGUMENTS ===
+# test-runner-runs: run1
+# test-runner-run/run1/app: ${ALL_CLUSTERS_APP}
+# test-runner-run/run1/factoryreset: True
+# test-runner-run/run1/quiet: True
+# test-runner-run/run1/app-args: --discriminator 1234 --KVS kvs1 --trace-to json:${TRACE_APP}.json
+# test-runner-run/run1/script-args: --storage-path admin_storage.json --commissioning-method on-network --discriminator 1234 --passcode 20202021 --endpoint 1 --trace-to json:${TRACE_TEST_JSON}.json --trace-to perfetto:${TRACE_TEST_PERFETTO}.perfetto
+# === END CI TEST ARGUMENTS ===
+
+import copy
+import logging
+import random
+
+import chip.clusters as Clusters
+from chip import ChipDeviceCtrl  # Needed before chip.FabricAdmin
+from chip.clusters import Globals
+from chip.clusters.Types import NullValue
+from chip.interaction_model import InteractionModelError, Status
+from matter_testing_support import MatterBaseTest, TestStep, async_test_body, default_matter_test_main
+from mobly import asserts
+
+logger = logging.getLogger(__name__)
+
+cluster = Clusters.Thermostat
+
+
+class TC_TSTAT_2_2(MatterBaseTest):
+
+    def desc_TC_TSTAT_2_2(self) -> str:
+        """Returns a description of this test"""
+        return "42.2.2. [TC-TSTAT-2.2] Setpoint Test Cases with server as DUT"
+
+    def pics_TC_TSTAT_2_2(self):
+        """This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return [self.check_pics("TSTAT.S")]
+
+    def steps_TC_TSTAT_2_2(self) -> list[TestStep]:
+        steps = [
+            TestStep("1", "Commission DUT to TH"),
+            TestStep("2a", "Test Harness Client reads  attribute OccupiedCoolingSetpoint from the DUT"),
```

Comment:
Can we store all initial attribute values (mainly min and max values) so that we can restore them at the end of the test? I wasn't able to run this test back to back since by the end of the test some of the attribute values don't make sense anymore.

Reply:
The exact issue I saw here is that, by the end of the test, the heatingsetpoint is very close to the abs/maxheatingsetpoint and some tests increment fixed values to heatingsetpoint which would be out of range and fail a test.

Reply:
This turned out to be harder than you might expect, as the default values for the thermostat cluster violate its own constraints. Once [PR 36023](https://github.com/project-chip/connectedhomeip/pull/36023) is merged, we can update the code to match, and then this will work.